ReSharper.FSharp/src/FSharp.Psi.Services/src/Util/SignatureFileUtil.fs

```diff
@@ -0,0 +1,172 @@
+                t.TypeDefinition.DisplayName
+            else
+                let isPostFix =
+                    set [| "list"; "option"; "array" |]
+                    |> Set.contains t.TypeDefinition.DisplayName
+                    
+                if isPostFix then
+                    let ga = t.GenericArguments[0].GenericParameter
+                    let tick = if ga.IsSolveAtCompileTime then "^" else "'"
+                    $"{tick}{ga.DisplayName} {t.TypeDefinition.DisplayName}"
+                else
+                    let args = Seq.map getTypeName t.GenericArguments |> String.concat ","
+                    $"{t.TypeDefinition.DisplayName}<{args}>"
+        elif t.IsGenericParameter then
+            let tick = if t.GenericParameter.IsSolveAtCompileTime then "^" else "'"
+            $"{tick}{t.GenericParameter.DisplayName}"
+        elif t.IsFunctionType then
+            let rec visit (t: FSharpType) : string seq =
+                if  t.IsFunctionType then
+                    Seq.collect visit t.GenericArguments
+                else
+                    Seq.singleton (getTypeName t)
+            
+            visit t
+            |> String.concat " -> "
+            |> sprintf "(%s)"
+        elif t.IsTupleType then
+            t.GenericArguments
+            |> Seq.map getTypeName
+            |> String.concat " * "
+        else
+            "???"
+
+    let rec visitAccessPath (t: FSharpType) : string list =
+        [
+            if t.HasTypeDefinition then
+                yield t.TypeDefinition.AccessPath
+            if not t.IsGenericParameter && not (Seq.isEmpty t.GenericArguments) then
+                yield! (Seq.collect visitAccessPath t.GenericArguments)
+        ]
+
+    let alwaysOpen =
+        let existingOpenStatements =
+            sigDecl.Members
+            |> Seq.choose (function | :? IOpenStatement as os -> Some os | _ -> None)
+            |> Seq.map (fun openStatement -> openStatement.ReferenceName.QualifiedName)
+
+        set
+            [|
+                yield "Microsoft.FSharp.Core"
+                yield "Microsoft.FSharp.Collections"
+                yield sigDecl.CompiledName
+                yield! existingOpenStatements
+            |]
+    
+    let openStatements =
+        List.collect visitAccessPath types
+        |> List.filter (fun p -> Set.contains p alwaysOpen |> not)
+        |> List.distinct
+        |> List.partition (fun p -> p.StartsWith("System"))
+        |> fun (systemPaths, otherPaths) ->
+            [ yield! List.sort systemPaths
+              yield! otherPaths ]
+        |> List.map elementFactory.CreateOpenStatement
+
+    let namedParameters =
+        Seq.zip types parameters
+        |> Seq.map (fun (t, p) ->
+            let typeName = getTypeName t
+            match p.Pattern.IgnoreInnerParens() with
+            | DeclaredNameInPattern name -> $"{name}: {typeName}"
+            | :? ITuplePat as tuplePat ->
```

Comment:
This may break with nested parens. We should check if, for example, in this code the two tuples have the same representation in signatures and compiled code or not:
```fsharp
let f (a, b) ((c, d)) = ...
```